---
title: "Microbiome Data Analysis workflow"
author: "Domenico Palumbo"
date: "11/19/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is an R Markdown document about an easy workflow for microbiome data analysis. This is a reduced version of "Workflow for Microbiome Data Analysis: from raw reads to community analyses" that you can find here: "https://bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/MicrobiomeWorkflowII.html". I also added other small things that you can find usefull.
I'll try to show you the basic procedures to obtain results from some test samples from MiSeq_SOP Dataset ("http://www.mothur.org/w/images/d/d6/MiSeqSOPData.zip").

Let's start loading the libraries

```{r libraries, echo=TRUE, message=F, warning=F}
library(ggplot2)
library(dada2)
library(vegan)
```

Then we'll download the test data

```{r test_data, message=FALSE, warning=FALSE}

miseq_path <- "~/MiSeq_SOP" #CHANGE DIRECTORY DEPENDING ON THE NEEDS

if(!file_test("-d", miseq_path)) {
     dir.create(miseq_path)
}
download.file("http://www.mothur.org/w/images/d/d6/MiSeqSOPData.zip", destfile = paste(miseq_path,"MiSeqSOPData.zip", sep ="/"))
unzip(paste(miseq_path,"MiSeqSOPData.zip", sep ="/"), exdir = miseq_path)
invisible(file.remove(file.path(miseq_path, "MiSeqSOPData.zip")))
unlink(x = paste(miseq_path,"__MACOSX", sep ="/"), recursive = T)
```

Sort ensures forward/reverse reads are in same order

```{r fastq, echo=T}
miseq_path2=paste(miseq_path,"MiSeq_SOP", sep ="/")
fnFs <- sort(list.files(miseq_path2, pattern="_R1_001.fastq"))
fnRs <- sort(list.files(miseq_path2, pattern="_R2_001.fastq"))
```

Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq

```{r sample_names, echo=T}
sampleNames <- sapply(strsplit(fnFs, "_"), `[`, 1)
```

Specify the full path to the fnFs and fnRs

```{r paths, echo=T}
fnFs <- file.path(miseq_path2, fnFs)
fnRs <- file.path(miseq_path2, fnRs)
```

These are two plots showing how our *.fastq look like:

```{r, include=TRUE, message=F}
plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])
```

**Trimming**

As you can see the quality decrease in the end of the reverse reads and in the beginning of all the reads. This is normal, but we decide to trim the reads to obtain better quality.

```{r trimming, echo=T}
filt_path <- file.path(miseq_path, "filtered") # Place filtered files in filtered/ subdirectory
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sampleNames, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sampleNames, "_R_filt.fastq.gz"))

#Filter the forward and reverse reads:
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160), maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, compress=TRUE, multithread=F) # On Windows set multithread=FALSE
```

**Dereplication**

After trimming, we can remove unwanted replicates with a dereplication process. We'll use DEREP

```{r derep1, echo=T,message=F}
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)

# Name the derep-class objects by the sample names
names(derepFs) <- sampleNames
names(derepRs) <- sampleNames
```

Dereplication combines all identical sequencing reads into into “unique sequences” with a corresponding “abundance”: the number of reads with that unique sequence. Dereplication substantially reduces computation time by eliminating redundant comparisons.

The DADA2 method relies on a parameterized model of substitution errors to distinguish sequencing errors from real biological variation. Because error rates can (and often do) vary substantially between sequencing runs and PCR protocols, the model parameters can be discovered from the data itself using a form of unsupervised learning in which sample inference is alternated with parameter estimation until both are jointly consistent.

Parameter learning is computationally intensive, as it requires multiple iterations of the sequence inference algorithm, and therefore it is often useful to estimate the error rates from a (sufficiently large) subset of the data.

```{r derep2, echo=T, message=T}
errF <- learnErrors(filtFs, multithread=F)
errR <- learnErrors(filtRs, multithread=F)
```

Now, we'll calculate the duplication level and we'll create the sequence table (similar to the OTU Table)

```{r dada, echo=T}
#We calculate the duplication
dadaFs <- dada(derepFs, err=errF, multithread=F)
dadaRs <- dada(derepRs, err=errR, multithread=F)

#We merge the files
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs)

#We create the table with the sequences 
seqtabAll <- makeSequenceTable(mergers[!grepl("Mock", names(mergers))]) #Similar to OTU Table
table(nchar(getSequences(seqtabAll)))
```

**Chimera removing**

```{r chimeras, echo=T, message=F}
#Remove chimeras
seqtabNoC <- removeBimeraDenovo(seqtabAll)
```

**Assign Taxonomy**

Now it's time to assign our sequences to a specific taxa. We do that using a reference genome for 16S sequences. We'll associate each sequence with a specific bacterium (a specific taxa) and we'll calculate also the abundace for each one in our samples.
First, we'll download the reference. The most used ones are: RDP, GreenGenes and Silva. In this case, we'll use an old version of RDP.

```{r taxonomy, echo=T, message=F}
#First, we'll download the reference
download.file(url = "https://github.com/BioH4z/Bioc_microbiome_analysis/raw/master/rdp_train_set_16.fa", destfile = "~/rdp_train_set_16.fa")
fastaRef <- "~/rdp_train_set_16.fa"

#Assigning the taxonomy
taxTab <- assignTaxonomy(seqtabNoC, refFasta = fastaRef, multithread=F)
```

**Alpha Diversity**

Alpha diversity refers to the average species diversity in a habitat or specific area and it is a local measure. 

Starting from the alpha diversity, we can consider three main index:

-Species richness (OTU count)  "How many?"
How many different species could be detected in a microbial ecosystem?

-Shannon index  "How different?"
How are the microbes balanced to each other? Do we have species evenness (similar abundance level) or do some species dominate others? Shannon index measures how evenly the microbes are distributed in a sample.

-Simpson's Diversity Index 
It is a measure of diversity which takes into account the number of species present, as well as the relative abundance of each species. As species richness and evenness increase, so diversity increases. The value of D ranges between 0 and 1.

```{r alpha, echo=T, include=T}
shannon=diversity(seqtabNoC, index = "shannon")
simpson=diversity(seqtabNoC, index = "simpson")

#Plots
par(mfrow = c(1, 2))
hist(simpson)
hist(shannon)
```


**Beta diversity**

Beta diversity refers to the ratio between local or alpha diversity and regional diversity. This is the diversity of species between two habitats or regions.

Next we can calcuate a pair-wise dissimilarity (distance) measure between sites based on their species composition using the function vegdist. Vegdist computes dissimilarity indices. We are using bray-curtis which is good in detecting underlying ecological gradients

It's used to quantify the compositional dissimilarity between two different sites. It is bounded between 0 and 1, where 0 = same composition, 1 = maximally dissimilar.

Dissimilarity analysis is a good way to explore variability in community composition. 

```{r beta, echo=T, include=T}
bray = vegdist(seqtabNoC, "bray") #You can choose different methods to calculate the distance. This is one of the most used but if you want to use another software instead of R you can find other methods.

hist(bray, xlim = range(0.0,1.0))
```

```{r bray-curtis plot, echo=T, message=F}
spe.b.pcoa <- cmdscale(bray, k=2)
ordiplot(scores(spe.b.pcoa, choices=c(1,2)), type="t", main="PCoA with species weighted averages")
```


**Rarefaction**

The issue that occurs when sampling various species in a community is that the larger the number of individuals sampled, the more species that will be found.
In ecology, rarefaction is a technique to assess species richness from the results of sampling. Rarefaction allows the calculation of species richness for a given number of individual samples, based on the construction of so-called rarefaction curves. This curve is a plot of the number of species as a function of the number of samples. Rarefaction curves generally grow rapidly at first, as the most common species are found, but the curves plateau as only the rarest species remain to be sampled. Rarefaction curves are necessary for estimating species richness. Raw species richness counts, which are used to create accumulation curves, can only be compared when the species richness has reached a clear asymptote. Rarefaction curves produce smoother lines that facilitate point-to-point or full dataset comparisons.

Rarefaction curves are created by randomly re-sampling the pool of N samples multiple times and then plotting the average number of species found in each sample (1,2, … N).
To try some rarefaction, we use the rarefy and rarecurve functions.

```{r rarefy, echo=T, include=T}
spAbund <- rowSums(seqtabNoC) #gives the number of individuals found in each plot
spAbund # view observations per plot

raremin <- min(rowSums(seqtabNoC)) #rarefaction uses the smallest number of observations per sample to extrapolate the expected number if all other samples only had that number of observations
raremin

sRare <- rarefy(seqtabNoC, raremin) # now use function rarefy
sRare #gives an "expected"rarefied" number of species (not obs) if only 2518 individuals were present

rarecurve(seqtabNoC, col = "blue") # produces rarefaction curves # squares are the samples' name at observed space. To "rarefy" a larger site, follow the rarefaction curve until the curve corresponds with the lesser site obs. This gives you rarefied species richness
```



